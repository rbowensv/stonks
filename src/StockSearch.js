import React, {useState} from 'react';
import { STOCK_SYMBOL } from './app-constants';

export function StockSearch(props) {
    const [searchText, setSearchText] = useState('');
    const clearSearch = () => {
        setSearchText('');
        clearSearchSuggestions();
    }

    const {
        searchSuggestions,
        getSearchSuggestion,
        pushStock,
        clearSearchSuggestions,
        viewedStocks
    } = props;

    return (
        <div className='stock-search'>
        {(viewedStocks.length < 3) && <div>
            <label htmlFor='stock-search' className='stock-search--label'>Search stocks</label>
            <input
            type='text'
            id='stock-search' 
            placeholder='Search...'
            value={searchText}
            className='stock-search--input' 
            onChange={(e) => {
                getSearchSuggestion(e.target.value)
                setSearchText(e.target.value)
                }} 
            />
            <div className='stock-search--suggestions'>
                { searchSuggestions && 
                    (searchSuggestions.map((result) => {
                    return <button 
                    key={result[STOCK_SYMBOL]}
                    value={result.symbol}
                    name={result.name}
                    className={`stock-search--suggestion ${result.symbol}`} 
                    onClick={(e) => {
                        pushStock({symbol: e.target.value, name: e.target.name})
                        setSearchText('');
                        clearSearch()
                        }}>
                        {result.displayName}
                    </button>
                }))}
            </div>
        </div>}
        {(viewedStocks.length === 3) && <div className='stock-search--full'>Remove a stock to compare a different one.</div>}
    </div>
    );
}