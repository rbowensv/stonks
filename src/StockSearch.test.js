import renderer, {act} from 'react-test-renderer'
import { StockSearch } from './StockSearch'

const mockPush = jest.fn();
const mockGet = jest.fn();
const mockClear = jest.fn();

const mockSetState = jest.fn();

// jest.mock('react', () => ({
//   useState: initial => [initial, mockSetState]
// }));

const mockSearchSuggestions = [
    {
        '1. symbol': "E",
        '2. name': "Eni Spa",
        '3. type': "Equity",
        '4. region': "United States",
        '5. marketOpen': "09:30",
        '6. marketClose': "16:00",
        '7. timezone': "UTC-04",
        '8. currency': "USD",
        '9. matchScore': "1.0000",
        displayName: "E: Eni Spa",
        name: "Eni Spa",
        symbol: "E"
    },
    {
        '1. symbol': "E.TRT",
        '2. name': "Enterprise Group",
        '3. type': "Equity",
        '4. region': "Toronto",
        '5. marketOpen': "09:30",
        '6. marketClose': "16:00",
        '7. timezone': "UTC-05",
        '8. currency': "CAD",
        '9. matchScore': "0.4000",
        'displayName': "E.TRT: Enterprise Group",
        name: "Enterprise Group",
        symbol: "E.TRT"
    }
]

const mockViewedStocks = []

const fakeClick = {
    target: {
        value: 'E',
        name: 'Eni Spa'
    }
}
const fakeType = {
    target: {
        value: 'test'
    }
}

test('renders as expected', () => {
    const component = renderer.create(<StockSearch
         searchSuggestions={mockSearchSuggestions} 
         getSearchSuggestion={mockGet} 
         pushStock={mockPush} 
         clearSearchSuggestions={mockClear}
         viewedStocks={mockViewedStocks}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

test('clicking on suggestion calls push stock and clear', () => {
    const component = renderer.create(<StockSearch
        searchSuggestions={mockSearchSuggestions} 
        getSearchSuggestion={mockGet} 
        pushStock={mockPush} 
        clearSearchSuggestions={mockClear}
        viewedStocks={mockViewedStocks}/>);
    let buttons = component.root.findAllByType("button");
    buttons[0].props.onClick(fakeClick);
    expect(mockPush).toHaveBeenCalled();
    expect(mockClear).toHaveBeenCalled();
})

//its mad because I haven't set up a place forstateful updating
test('typing in the search box calls get', () => {
    const component = renderer.create(<StockSearch
        searchSuggestions={mockSearchSuggestions} 
        getSearchSuggestion={mockGet} 
        pushStock={mockPush} 
        clearSearchSuggestions={mockClear}
        viewedStocks={mockViewedStocks}/>);
    let input = component.root.findByType('input');
    act(() => {
        input.props.onChange(fakeType);
    })
    expect(mockGet).toHaveBeenCalled();
})