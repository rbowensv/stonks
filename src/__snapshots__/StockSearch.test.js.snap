// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div
  className="stock-search"
>
  <div>
    <label
      className="stock-search--label"
      htmlFor="stock-search"
    >
      Search stocks
    </label>
    <input
      className="stock-search--input"
      id="stock-search"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
    <div
      className="stock-search--suggestions"
    >
      <button
        className="stock-search--suggestion E"
        name="Eni Spa"
        onClick={[Function]}
        value="E"
      >
        E: Eni Spa
      </button>
      <button
        className="stock-search--suggestion E.TRT"
        name="Enterprise Group"
        onClick={[Function]}
        value="E.TRT"
      >
        E.TRT: Enterprise Group
      </button>
    </div>
  </div>
</div>
`;
